### DISCUSSION TITLE
Cannot save HardSamplingClassifier

### DISCUSSION THREAD

**Discussion by nehalAggarwal** [online-ml/river#544] [Q&A] [answered]
Hello, 

I have run models using HardSamplingClassifier and I wished to save the model using pickle or joblib. 

It seems that the model is not ready to be saved as it gives Pickling Error as - 
![hardSample](https://user-images.githubusercontent.com/20676350/112770122-18278480-8ff3-11eb-835f-451ad5750f32.PNG)

Is there a way I can fix this error and save/pickle HardSamplingClassifier model?

**Comment by raphaelsty** [comment, marked as answer]
Hi @nehalAggarwal, 

Pickle does not allow serializing lambda function. The `HardSampling` class is one of the few classes in River to have a lambda function in it's source code.  I suggest you to use the [Dill](https://github.com/uqfoundation/dill) library to serialize this model. Dill is more or less the same as Pickle, maybe a bit slower but allows to serialize lambda functions.

You can install Dill using:

```sh
pip install dill
```

```python
from river import datasets
from river import evaluate
from river import imblearn
from river import linear_model
from river import metrics
from river import optim
from river import preprocessing

model = (
 preprocessing.StandardScaler() |
 imblearn.HardSamplingClassifier(
     classifier=linear_model.LogisticRegression(),
     p=0.1,
     size=40,
     seed=42,
 )
)
```

```python
import dill

with open('./model.pkl', 'wb') as output_file:
    dill.dump(model, output_file)
```

```python
import dill

with open('./model.pkl', 'rb') as input_file:
    model = dill.load(input_file)
```

**Reply by MaxHalford** [reply]
@raphaelsty is right, and yet we could and should fix `HardSamplingClassifier` so that it doesn't use a `lambda` anymore.

**Reply by MaxHalford** [reply]
This is fixed in #545.
